core/src/main/java/com/microsoft/applicationinsights/internal/heartbeat/HeartBeatProvider.java

```diff
@@ -0,0 +1,304 @@
+          setResult = false;
+          throw new Exception("heartbeat beat property specified is a reserved property");
+        }
+        else  {
+          HeartBeatPropertyPayload payload = new HeartBeatPropertyPayload();
+          payload.setHealthy(isHealthy);
+          payload.setPayloadValue(propertyValue);
+          heartbeatProperties.put(propertyName, payload);
+          setResult = true;
+        }
+      }
+      catch (Exception e) {
+        InternalLogger.INSTANCE.warn("failed to set heartbeat property name %s, value %s, "
+            + "stack trace is: %s", propertyName, propertyValue, ExceptionUtils.getStackTrace(e));
+      }
+    }
+    else {
+      InternalLogger.INSTANCE.warn("cannot set property without property name");
+    }
+    return setResult;
+  }
+
+  @Override
+  public boolean isHeartBeatEnabled() {
+    return isEnabled;
+  }
+
+  @Override
+  public void setHeartBeatEnabled(boolean isEnabled) {
+    this.isEnabled = isEnabled;
+  }
+
+  @Override
+  public List<String> getExcludedHeartBeatPropertyProviders() {
+    return this.disabledHeartBeatPropertiesProviders;
+  }
+
+  @Override
+  public void setExcludedHeartBeatPropertyProviders(
+      List<String> excludedHeartBeatPropertyProviders) {
+    this.disabledHeartBeatPropertiesProviders = excludedHeartBeatPropertyProviders;
+  }
+
+  @Override
+  public long getHeartBeatInterval() {
+    return this.interval;
+  }
+
+  @Override
+  public void setHeartBeatInterval(long timeUnit) {
+    // user set time unit in seconds
+    if (timeUnit <= HeartBeatProviderInterface.MINIMUM_HEARTBEAT_INTERVAL) {
+      this.interval = HeartBeatProviderInterface.MINIMUM_HEARTBEAT_INTERVAL;
+    }
+    else {
+      this.interval = timeUnit;
+    }
+  }
+
+  @Override
+  public List<String> getExcludedHeartBeatProperties() {
+    return this.disableDefaultProperties;
+  }
+
+  @Override
+  public void setExcludedHeartBeatProperties(List<String> excludedHeartBeatProperties) {
+    this.disableDefaultProperties = excludedHeartBeatProperties;
+  }
+
+  @Override
+  public void stop(long timeout, TimeUnit timeUnit) {
+    propertyUpdateService.shutdown();
+    heartBeatSenderService.shutdown();
+    try {
+      propertyUpdateService.awaitTermination(timeout, timeUnit);
+      heartBeatSenderService.awaitTermination(timeout, timeUnit);
+    }
+    catch (InterruptedException e) {
+      InternalLogger.INSTANCE.warn("unable to successfully terminate heartbeat module, "
+          + "encountered and exception with stacktrace, %s", ExceptionUtils.getStackTrace(e));
+    }
+  }
+
+  /**
+   * Send the heartbeat item synchronously to application insights backend.
+   */
+  private void send() {
+
+    synchronized (lock) {
+      MetricTelemetry telemetry = (MetricTelemetry)gatherData();
+      telemetry.getContext().getOperation().setSyntheticSource(HEARTBEAT_SYNTHETIC_METRIC_NAME);
+      telemetryClient.trackMetric(telemetry);
+      heartbeatsSent.incrementAndGet();
```

Comment:
This can be in the log line below. It returns the value after incrementing. Also, since it's in a lock it doesn't need to be atomic

Reply:
See other comment regarding the lock. If you remove the lock, you can keep the atomic int.

Reply:
Fixed